src/boot/efi/stub.c

```diff
@@ -180,6 +181,139 @@ static bool use_load_options(
+        if (err == EFI_NOT_FOUND)
+                /* No extra subdir, that's totally OK */
+                goto nothing;
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to open addons directory: %m");
+
+        err = get_file_info(extra_dir, &extra_dir_info, NULL);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to query information of addons directory: %m");
+
+        for (;;) {
+                _cleanup_free_ char16_t *d = NULL;
+
+                err = readdir(extra_dir, &dirent, &dirent_size);
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to read extra directory of loaded image: %m");
+                if (!dirent) /* End of directory */
+                        break;
+
+                if (dirent->FileName[0] == '.')
+                        continue;
+                if (FLAGS_SET(dirent->Attribute, EFI_FILE_DIRECTORY))
+                        continue;
+                if (!is_ascii(dirent->FileName))
+                        continue;
+                if (strlen16(dirent->FileName) > 255) /* Max filename size on Linux */
+                        continue;
+
+                d = xstrdup16(dirent->FileName);
+
+                if (n_items+2 > n_allocated) {
+                        /* We allocate 16 entries at a time, as a matter of optimization */
+                        if (n_items > (SIZE_MAX / sizeof(uint16_t)) - 16) /* Overflow check, just in case */
+                                return log_oom();
+
+                        size_t m = n_items + 16;
+                        items = xrealloc(items, n_allocated * sizeof(uint16_t *), m * sizeof(uint16_t *));
+                        n_allocated = m;
+                }
+
+                items[n_items++] = TAKE_PTR(d);
+                items[n_items] = NULL; /* Let's always NUL terminate, to make freeing via strv_free() easy */
+        }
+
+        if (n_items == 0)
+                /* Empty directory */
+                goto nothing;
+
+        /* Now, sort the files we found, to make this uniform and stable (and to ensure the TPM measurements
+         * are not dependent on read order) */
+        sort_pointer_array((void**) items, n_items, (compare_pointer_func_t) strcmp16);
+
+        for (size_t i = 0; i < n_items; i++) {
+                _cleanup_free_ EFI_DEVICE_PATH *addon_path = NULL;
+                _cleanup_(unload_imagep) EFI_HANDLE addon = NULL;
+                _cleanup_free_ char16_t *addon_spath = NULL;
+                EFI_LOADED_IMAGE_PROTOCOL *loaded_addon = NULL;
+                size_t section_offset = 0, section_size = 0;  /* avoid false maybe-uninitialized warning */
+
+                static const char * const sections[2] = {
+                        [0] = ".cmdline",
+                        NULL,
+                };
+
+                addon_spath = xasprintf("\\loader\\%ls\\%ls", extra_dir_info->FileName, items[i]);
```

Comment:
Instead of formatting a full path and operating by that, cant we just open the thing right-away in the loop above and keep it open? I.e. openat() style it sounds bad to me to drop the rer to the dir and traverse the path afresh for each dropin

Reply:
We need to go through LoadImage for the verification protocol, I don't think it takes open handles?